cluster-autoscaler/core/scale_down.go

```diff
@@ -342,35 +346,85 @@ func getEmptyNodes(candidates []*apiv1.Node, pods []*apiv1.Pod, maxEmptyBulkDele
 }
 
+func evictPod(podToEvict *apiv1.Pod, client kube_client.Interface, recorder kube_record.EventRecorder,
+	maxGratefulTerminationSec int, retryUntil time.Time, waitBetweenRetries time.Duration) error {
+	recorder.Eventf(podToEvict, apiv1.EventTypeNormal, "ScaleDown", "deleting pod for node scale down")
+	maxGraceful64 := int64(maxGratefulTerminationSec)
+	var lastError error
+	first := true
+	for first || time.Now().Before(retryUntil) {
+		first = false
+		eviction := &policyv1.Eviction{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: podToEvict.Namespace,
+				Name:      podToEvict.Name,
+			},
+			DeleteOptions: &metav1.DeleteOptions{
+				GracePeriodSeconds: &maxGraceful64,
+			},
+		}
+		lastError = client.Core().Pods(podToEvict.Namespace).Evict(eviction)
+		if lastError == nil {
+			return nil
+		}
+		time.Sleep(waitBetweenRetries)
+	}
+	glog.Errorf("Failed to evict pod %s, error: %v", podToEvict.Name, lastError)
+	recorder.Eventf(podToEvict, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to delete pod for ScaleDown")
+	return fmt.Errorf("Failed to evict pod %s/%s within allowed timeout (last error: %v)", podToEvict.Namespace, podToEvict.Name, lastError)
+}
+
 // Performs drain logic on the node. Marks the node as unschedulable and later removes all pods, giving
 // them up to MaxGracefulTerminationTime to finish.
 func drainNode(node *apiv1.Node, pods []*apiv1.Pod, client kube_client.Interface, recorder kube_record.EventRecorder,
-	maxGratefulTerminationSec int) error {
+	maxGratefulTerminationSec int, maxPodEvictionTime time.Duration, waitBetweenRetries time.Duration) error {
+
+	drainSuccessfull := false
+	toEvict := len(pods)
 	if err := deletetaint.MarkToBeDeleted(node, client); err != nil {
 		recorder.Eventf(node, apiv1.EventTypeWarning, "ScaleDown", "failed to mark the node as toBeDeleted/unschedulable: %v", err)
 		return err
 	}
+
+	// If we fail to evict all the pods from the node we want to remove delete taint
+	defer func() {
+		if !drainSuccessfull {
+			deletetaint.CleanToBeDeleted(node, client)
+			recorder.Eventf(node, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to drain the node, aborting ScaleDown")
+		}
+	}()
+
 	recorder.Eventf(node, apiv1.EventTypeNormal, "ScaleDown", "marked the node as toBeDeleted/unschedulable")
 
-	maxGraceful64 := int64(maxGratefulTerminationSec)
+	retryUntil := time.Now().Add(maxPodEvictionTime)
+	confirmations := make(chan error, toEvict)
 	for _, pod := range pods {
-		recorder.Eventf(pod, apiv1.EventTypeNormal, "ScaleDown", "deleting pod for node scale down")
-		err := client.Core().Pods(pod.Namespace).Delete(pod.Name, &metav1.DeleteOptions{
-			GracePeriodSeconds: &maxGraceful64,
-		})
+		go func(podToEvict *apiv1.Pod) {
+			confirmations <- evictPod(podToEvict, client, recorder, maxGratefulTerminationSec, retryUntil, waitBetweenRetries)
+		}(pod)
+	}
+
+	evictionErrs := make([]error, 0)
+
+	for range pods {
+		err := <-confirmations
```

Comment:
In theory this can hung indefinitely. You try to retry until maxPodEvictionTime) but it may happen that the send takes forever. 

select {
    case <- time.After(retryUntil.Sub(time.Now) + 5 * time.sec): 
       glog.Errorf("WARNING - SOMETHING STUCKED  UP - should have stopped all pod evictions long time ago) 
       evictionErrors = append(evictionErrors, something stucked up)
   case err := <- contirmations
     ....
}